ex/app/src/main/java/com/example/ex/MainFragment.java

```diff
@@ -0,0 +1,209 @@
+package com.example.ex;
+
+import android.content.Context;
+import android.os.Bundle;
+import androidx.annotation.NonNull;
+import androidx.annotation.Nullable;
+import androidx.fragment.app.Fragment;
+import androidx.fragment.app.FragmentActivity;
+import androidx.lifecycle.Observer;
+import androidx.lifecycle.ViewModelProvider;
+import androidx.recyclerview.widget.LinearLayoutManager;
+import androidx.recyclerview.widget.RecyclerView;
+import android.os.Handler;
+import android.util.Log;
+import android.view.LayoutInflater;
+import android.view.View;
+import android.view.ViewGroup;
+import android.widget.Button;
+import android.widget.EditText;
+import android.widget.LinearLayout;
+import android.widget.ProgressBar;
+import android.widget.Toast;
+
+import java.util.ArrayList;
+
+public class MainFragment extends Fragment {
+
+
+    private static final int UPPER_BOUND = 1000000;
+    private static final int DEFAULT_SLEEP_TIME = 1000;
+    private static final int HUNDRED = 100;
+    private MainViewModel viewModel;
+    private RecyclerView recyclerView;
+    private RecyclerViewAdapter recyclerViewAdapter;
+    private EditText editText;
+    private IAttachable listener;
+    private ProgressBar progressBar;
```

Comment:
There was a certain comment regarding `Fragment`'s `View`s leaks in our `ViewModel and LiveData` guide in the `Note on Fragment Views leaks` section. Please check it out